src/main/java/uk/gov/hmcts/reform/sandl/snlevents/service/HearingService.java

```diff
@@ -0,0 +1,200 @@
+
+@Service
+public class HearingService {
+
+    @Autowired
+    HearingRepository hearingRepository;
+
+    @Autowired
+    private EntityManager entityManager;
+
+    public Page<HearingSearchResponse> search(List<SearchCriteria> searchCriteriaList, Pageable pageable) {
+
+        CriteriaBuilder cb = entityManager.getCriteriaBuilder();
+        CriteriaQuery<HearingSearchResponse> cq = cb.createQuery(HearingSearchResponse.class);
+        Root<Hearing> hearingRoot = cq.from(Hearing.class); //root entity
+
+        //Subquery
+        Subquery<String> subqueryPerson = cq.subquery(String.class);
+        Root<Person> subqueryPersonRoot = subqueryPerson.from(Person.class);
+        subqueryPerson.where(
+            //join subquery with main query
+            cb.equal(subqueryPersonRoot.get("id"), hearingRoot)
+        );
+        subqueryPerson.select(cb.greatest(subqueryPersonRoot.<String>get("name")));
+
+
+        //Subquery
+        Subquery<Long> subqueryListingCount = cq.subquery(Long.class);
+        Root<HearingPart> subqueryListingCountRoot = subqueryListingCount.from(HearingPart.class);
+        subqueryListingCount.select(cb.count(subqueryListingCountRoot));
+        subqueryListingCount.where(
+            //join subquery with main query
+            cb.equal(subqueryListingCountRoot.get(HearingPart_.hearingId), hearingRoot),
+            // additional filters
+            cb.isNotNull(subqueryListingCountRoot.get(HearingPart_.sessionId))
+        );
+
+        //Subquery
+        Subquery<OffsetDateTime> subqueryListingStart = cq.subquery(OffsetDateTime.class);
+        Root<HearingPart> subqueryListingStartRoot = subqueryListingStart.from(HearingPart.class);
+        subqueryListingStart.where(
+            //join subquery with main query
+            cb.equal(subqueryListingStartRoot.get(HearingPart_.hearingId), hearingRoot),
+            // additional filters
+            cb.isNotNull(subqueryListingStartRoot.get(HearingPart_.sessionId))
+        );
+        subqueryListingStart.select(cb.least(subqueryListingStartRoot.<OffsetDateTime>get("start")));
+
+
+        Predicate restrictions = cb.conjunction();
+
+        for (SearchCriteria criteria: searchCriteriaList) {
+            ComparisonOperations operation = criteria.getOperation();
+            Predicate pred = null;
+            if (criteria.getKey().equals("listingStatus")
+                && operation.equals(ComparisonOperations.EQUALS)) {
+                boolean isListed = criteria.getValue().toString().equals("listed");
+
+                Subquery<HearingPart> subquery = cq.subquery(HearingPart.class);
+                Root<HearingPart> hpRoot = subquery.from(HearingPart.class);
+                subquery.select(hpRoot);
+                if (isListed) {
+                    subquery.where(cb.equal(hpRoot.get(HearingPart_.hearingId), hearingRoot),
+                        cb.isNotNull(hpRoot.get(HearingPart_.sessionId)));
+                } else {
```

Comment:
oh shit ;/

Reply:
@zielinskikamil wtf?

Reply:
I've just imagined that I need to change sth here ^^. Looks very busy but tbh I don't have a clue how we can split it.